cs-verifiable-encryption/README.md

```diff
@@ -0,0 +1,136 @@
+This verifiable encryption is meant to be treated as a predicate.  
+The next section will describe the API of this protocol.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This protocol has been implemented in this [PR](https://github.com/hyperledger/ursa/pull/40). The PR contains tests to demonstrate how the protocol works.
+
+1. The auditor can create his encryption keypair using `CSKeypair::new`. This method takes the maximum number of attributes that can be encrypted using the keypair. The following example creates a keypair whose public key can be used to encrypt at max 3 attributes. Trying to encrypt more will result in an error.  The `CSKeypair` is an struct of 2 fields, a public key `CSEncPubkey` and a private key `CSEncPrikey`.  
+
+    ```rust
+    let keypair: CSKeypair = CSKeypair::new(3)?;
+    let pub_key: &CSEncPubkey = &keypair.pub_key;
+    let pri_key: &CSEncPrikey = &keypair.pri_key;
+    ```
+
+1. Say the prover now wants to encrypt 3 attributes. He creates a sequence of his 2 attributes uses the above keypair's public key to call `CSKeypair::encrypt`. The `CSKeypair::encrypt` in addition to the public key takes a `label` which is a public value and indicates the condition underwhich the ciphertext should be decrypted. The same `label` should be used while decrypting this ciphetext. The `CSKeypair::encrypt` function returns a `CSCiphertext` if there is no error. 
+    
+    ```rust
+    let pub_key: &CSEncPubkey = &keypair.pub_key;
+    let attributes = vec!["Attribute1", "Attribute2", "Attribute3"];
+    let ciphertext: CSCiphertext = CSKeypair::encrypt(&attributes, "only_decrypt_when_*".as_bytes(), &pub_key)?;
+    ```
+
+1. To decrypt the ciphertext from above, the auditor will use the label, ciphertext and his keys in `CSKeypair::decrypt`. If successful, it returns the decrypted attributes
+
+    ```rust
+    let decrypted_messages: Vec<BigNumber> = CSKeypair::decrypt(label, &ciphertext, &keypair.pub_key, &keypair.pri_key)?;
+    ```
+
+1. The prover can now execute the verifiable encryption protocol. He will first either create or retrieve `blindings` for the attributes he wants to encrypt and then create the ciphertext over the attributes, over the `blindings` using `CSKeypair::encrypt_and_prove_phase_1`. He also gets some random values that he will use while creating the response to the challenge.
+
+    ```rust
+    let (ciphertext, blindings_ciphertext, r, r_tilde) = CSKeypair::encrypt_and_prove_phase_1(
+            &attributes,
+            &blindings,
+            label,
+            &keypair.pub_key,
```

Comment:
As above, probably better to reference (by way variable name and/or showing deserializing) the auditor pub_key than imply referencing a keypair member.

Reply:
Done.